ddsketch/src/lib.rs

```diff
@@ -0,0 +1,461 @@
+    fn bin_idx_to_store_idx(&mut self, bin_index: i32) -> usize {
+        if self.bins.is_empty() {
+            // If the bins are empty, start them at the index
+            self.offset = bin_index;
+            self.bins.push_back(0.0);
+            return 0;
+        }
+
+        // General case
+        // Bucket lower than the stored range
+        if bin_index < self.offset {
+            let additional_low_bins = self.offset - bin_index;
+            debug_assert!(additional_low_bins >= 0);
+
+            let additional_low_bins = std::cmp::min(
+                additional_low_bins as usize,
+                self.max_size as usize - self.bins.len(),
+            );
+
+            self.bins.reserve(additional_low_bins);
+            for _ in 0..additional_low_bins {
+                self.bins.push_front(0.0);
+            }
+
+            self.offset -= additional_low_bins as i32;
+            #[allow(clippy::needless_return)]
+            return 0;
+        }
+        // Bucket higher than the stored range
+        else if self.offset + self.bins.len() as i32 <= bin_index {
+            let bin_range_size = bin_index - self.offset + 1; // Number of bucket to store
+
+            if bin_range_size > self.max_size {
+                self.collapse_low_bins(bin_range_size - self.max_size);
+            }
+            debug_assert!(self.bins.len() as i32 <= self.max_size);
+
+            let store_index = bin_index - self.offset;
+            for _ in 0..(store_index - self.bins.len() as i32 + 1) {
+                self.bins.push_back(0.0);
+            }
+            return store_index as usize;
+        }
+        // Bucket within the stored range
+        else {
+            return (bin_index - self.offset) as usize;
+        }
+    }
+
+    /// Collapse the `bin_number` lowest bins
+    fn collapse_low_bins(&mut self, bin_number: i32) {
+        let mut count = 0.0;
+        for _ in 0..bin_number {
+            count += self.bins.pop_front().unwrap_or(0.0);
+        }
+        if let Some(lowest_bin) = self.bins.front_mut() {
+            *lowest_bin += count;
+        } else {
+            self.bins.push_front(count);
+        }
+        self.offset += bin_number;
+    }
+}
+
+impl Default for LowCollapsingDenseStore {
+    fn default() -> Self {
+        Self::new(2048).unwrap()
+    }
+}
+
+/// Logarithmic mapping of bucket index to value
+#[derive(Debug, Clone, Copy)]
+struct LogMapping {
+    gamma: f64,
+    multiplier: f64,
+    min_indexable_value: f64,
+    index_offset: f64,
+}
+
+impl LogMapping {
+    fn new(gamma: f64, offset: f64) -> Option<Self> {
+        if gamma <= 1.0 {
+            return None;
+        }
+        let multiplier = Self::multiplier_from_gamma(gamma);
+        Some(Self {
+            gamma,
+            multiplier,
+            min_indexable_value: max(
+                // So that the value representing the lowest bucket is >= std::f64::MIN_POSITIVE
+                std::f64::MIN_POSITIVE * gamma,
```

Comment:
Just an FYI, it looks like `std::f64::MIN_POSITIVE` has been deprecated in favor of `f64::MIN_POSITIVE`.